### PR TITLE
Issue 6144: Health Checks for Controller Services 

### LINKED ISSUE #6144: Health Checks for Controller Services
**Is your feature request related to a problem? Please describe.**
This issue lists health checks to be implemented on Controller Health Service.
The purpose of these checks is to ensure all Controller services are up and running and healthy.
The Operators would use the /readiness and /liveness APIs on the Controller Health Service to know if the Controller is ready and alive respectively. This issue describes the checks that need to be performed as part of these API requests on Controller.

**Describe the solution you'd like**
The Health Service needs to report health of the Controller based on health information of its sub-component services.
Controller currently runs the following services (_in addition to the Health Service itself_) 
1. GRPC Server 
2. REST Server
3. SegmentContainerMonitor
4. Retention Service 
5. Watermarking Service 
6. Event Processor 
7. Cluster Listener

**Liveness check** for all services can be to check if the service is running using the `service.isRunning()` method.
If the liveness check passes for all sub-component services, Controller "/liveness" would return `true` else return `false`.
In addition to invoking this API, operators should also continue to do the current port checks for liveness.

Controller Readiness check would be a logical "and" of the readiness checks for each of its component services:
**Readiness check** for each component could be done as follows:
1. GRPC Server & 
--> isRunning() + ControllerEventProcessor ready.

2. Segment Container Monitor ---> 
isRunning() +
ClusterZKImpl  - check if ZK Session is healthy - can be done based on event returning "ERROR" State.

3. Cluster Listener  
isRunning() +
Add a new isReady() API to `ControllerClusterListener` that invokes isReady of each of the FailoverSweeper 
+ isZKConnected()

4. Retention Service &  Watermarking Service ---  isRunning() 
+ ZK Session ( ZKBucketStoreManager ---> ZKBucketStore --> ZKStoreHelper--> CuratorClient()
BucketService --> processNotification --> ConnectivityError  -- Error connecting to ZK

5. Event Processor 
-- change the current implementation of isReady() method as it just invokes `svc.isRunning()`
--- Add a boolean variable that changes to `true` when bootstrap completes.
--- Check if _system scope and streams used for doing event processing for the Streams API, txn processing, KVT API processing are present

### PR DESCRIPTION
**Pull request by anishakj**
Signed-off-by: anishakj <anisha.kj@dell.com>

**Change log description**  
Added liveness/readiness check for controller services

**Purpose of the change**  
Fixes #6144

**What the code does**  
For each of the services in controller  namely ,(GRPC, SegmentContainerMonitor, EventProcessors, Retention service, water marking service, cluster Listener) liveness check for each ensures the service is running. And  readiness check  performs some more additional checks.  Below are the checks added for each of the services

Segment Container Monitor- Check zookeeper service is connected. For  this added a new getter method isZKConnected()
Cluster Listener - Check zookeeper service is connected as well as all failover sweepers are ready. For this added a new method isReady()
Retention Service - Checks zookeeper service is connected, for this added a new method isZKConnected in zookeeperBucketManager
Watermarking service- Check zookeeper service is connected, for this added a new method isZKConnected in zookeeperBucketManager
EventProcessors - Check controller boot strapping is completed, For this a new instance variable isReady is added and will be set to true when bootstrap process is completed.

when each of these service passes liveness check will update status to `NEW` 
when each of these service passes readiness check will update status to `UP`

**How to verify it**  
Verified the health checks by running Pravega in standalone mode,  and liveness and readiness checks are passing using `curl -v http://localhost:9091/v1/health`

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java [hunk 1]
```diff
@@ -208,12 +211,7 @@ protected void shutDown() {
         }
     }
 
-    @Override
-    public boolean isReady() {
```
Comment (shrids) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java:212]: Any specific reason why this is removed?

File: controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import io.pravega.controller.fault.ControllerClusterListener;
+import io.pravega.shared.health.Health;
+import io.pravega.shared.health.Status;
+import io.pravega.shared.health.impl.AbstractHealthContributor;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ClusterListenerHealthContributor  extends AbstractHealthContributor {
+    private  ControllerClusterListener controllerClusterListener;
```
Comment (shrids) [inline controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java:26]: final here and below.

File: controller/src/main/java/io/pravega/controller/server/health/GRPCServerHealthContributor.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    public GRPCServerHealthContributor(String name, GRPCServer grpcServer) {
+        super(name);
+        this.grpcServer = grpcServer;
+    }
+
+    @Override
+    public Status doHealthCheck(Health.HealthBuilder builder) throws Exception {
+        boolean running = grpcServer.isRunning();
+        Status status = Status.DOWN;
+        if (running) {
```
Comment (shrids) [inline controller/src/main/java/io/pravega/controller/server/health/GRPCServerHealthContributor.java:36]: Could you elaborate on why `Status.New `is not used here?

File: shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java [hunk 4]
```diff
@@ -64,6 +65,9 @@
 
     private final CuratorFramework client;
 
+    @Getter
+    private boolean isZKConnected = true;
```
Comment (shrids) [inline shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java:69]: private final AtomicBoolean

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java (hunk 1 above)
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java:212]: isReady() is just invoking isRunning function. I have added a new instance variable `isReady` and a getter method. That is why removed this.

File: controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java (hunk 2 above)
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java:26]: will change it

File: controller/src/main/java/io/pravega/controller/server/health/GRPCServerHealthContributor.java (hunk 3 above)
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/health/GRPCServerHealthContributor.java:36]: For GRPC, there is no additional readiness checks. So if the liveness check is passing, making it as ready.

File: shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java (hunk 4 above)
Comment (anishakj) [reply shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java:69]: At line no:82, we are updating this variable to newState.isConnected() which is of type boolean.

Comment (anishakj) [reply shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java]: Modified to AtomicBoolean

File: controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    public Status doHealthCheck(Health.HealthBuilder builder) throws Exception {
+        boolean running = controllerClusterListener.isRunning();
+        Status status = Status.DOWN;
+        if (running) {
+            status = Status.NEW;
+        }
+        boolean ready = controllerClusterListener.isReady();
+        if (running && ready) {
+            status = Status.UP;
+
```
Comment (pbelgundi) [inline controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java:42]: nit: remove unnecessary newline.

File: shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java [hunk 6]
```diff
@@ -64,6 +65,8 @@
 
     private final CuratorFramework client;
 
+    private final AtomicBoolean isZKConnected = new AtomicBoolean(true);
```
Comment (pbelgundi) [inline shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java:68]: Initial value should be false.

File: shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java [hunk 7]
```diff
@@ -109,6 +114,10 @@ public void deregisterHost(Host host) {
         close(node);
     }
 
+    public boolean isZKConnected() {
+        return isZKConnected.get();
```
Comment (pbelgundi) [inline shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java:118]: Please add tests to cover this new code in all classes.

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java [hunk 8]
```diff
@@ -121,6 +122,8 @@
     private final long rebalanceIntervalMillis;
     private final AtomicLong truncationInterval;
     private ScheduledExecutorService rebalanceExecutor;
+    @Getter
+    private boolean isReady = false;
```
Comment (pbelgundi) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java:126]: We need a zookeeper connectivity check for ControllerEventProcessor as well, as this too depends on Zookeeper..... perhaps I missed mentioning this on the issue.... If `CheckpointStore` is of type `ZKCheckpointStore` we need to check connectivity to Zookeeper.

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: controller/src/main/java/io/pravega/controller/server/health/WatermarkingServiceHealthContributor.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Override
+    public Status doHealthCheck(Health.HealthBuilder builder) throws Exception {
+        boolean running = watermarkingService.isRunning();
+        boolean zkHealthy  = true;
+        Status status = Status.DOWN;
+        if (running) {
+            status = Status.NEW;
+        }
+        if (watermarkingService instanceof ZooKeeperBucketManager) {
```
Comment (eolivelli) [inline controller/src/main/java/io/pravega/controller/server/health/WatermarkingServiceHealthContributor.java:41]: is it possible to not use `instanceof` and add a method to BucketManager like "isMetadataServiceConnected".
usually `instanceof` is a code smell in these cases

Also I won't call it "isZKConnected" but something that is not directly related to ZK, at this level it is not important that we are dealing with ZK, it is an implementation detail

File: controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java [hunk 10]
```diff
@@ -67,6 +68,26 @@ public ControllerClusterListener(final Host host, final Cluster cluster,
         this.sweepers = Lists.newArrayList(sweepers);
     }
 
+    public boolean isZKConnected() {
+        if (cluster instanceof ClusterZKImpl) {
```
Comment (eolivelli) [inline controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java:72]: is it possible to not use `instanceof` and add a method to `io.pravega.common.cluster.Cluster` like "isMetadataServiceConnected".
usually `instanceof` is a code smell in these cases

Also I won't call it "isZKConnected" but something that is not directly related to ZK, at this level it is not important that we are dealing with ZK, it is an implementation detail

File: controller/src/main/java/io/pravega/controller/fault/SegmentContainerMonitor.java [hunk 11]
```diff
@@ -89,6 +95,10 @@ public SegmentContainerMonitor(HostControllerStore hostStore, CuratorFramework c
         );
     }
 
+    public boolean isZKConnected() {
```
[the remark this answers is not in the document]
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/fault/SegmentContainerMonitor.java]: this is of type AtomicBoolean, so cant use Lombok Getter

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java [hunk 12]
```diff
@@ -139,7 +142,7 @@ public ControllerEventProcessors(final String host,
     }
 
     @VisibleForTesting
-    ControllerEventProcessors(final String host,
+    public ControllerEventProcessors(final String host,
```
[the remark this answers is not in the document]
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java:146]: This is made public to use in the test of eventprocessorhealth contributor

File: controller/src/main/java/io/pravega/controller/server/health/RetentionServiceHealthContributor.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Override
+    public Status doHealthCheck(Health.HealthBuilder builder) throws Exception {
+        boolean running = retentionService.isRunning();
+        boolean zkHealthy = true;
+        Status status = Status.DOWN;
+        if (running) {
+            status = Status.NEW;
+        }
+        if (retentionService instanceof ZooKeeperBucketManager) {
```
[the remark this answers is not in the document]
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/health/RetentionServiceHealthContributor.java:43]: Fixed

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java (hunk 8 above)
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java]: Added that

File: controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+import io.pravega.controller.fault.ControllerClusterListener;
+import io.pravega.shared.health.Health;
+import io.pravega.shared.health.Status;
+import io.pravega.shared.health.impl.AbstractHealthContributor;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ClusterListenerHealthContributor  extends AbstractHealthContributor {
+    private final ControllerClusterListener controllerClusterListener;
+      public ClusterListenerHealthContributor(String name, ControllerClusterListener controllerClusterListener) {
```
Comment (shrids) [inline controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java:27]: @NotNull check here.

File: controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java [hunk 15]
```diff
@@ -341,10 +348,6 @@ protected void startUp() {
                 controllerClusterListener.startAsync();
             }
 
-            // Start the Health Service.
-            healthServiceManager = new HealthServiceManager(serviceConfig.getHealthCheckFrequency());
-            healthServiceManager.start();
```
[the remark this answers is not in the document]
Comment (shrids) [reply controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java:346]: Also, please add a line indicating that the healthServiceManager has started or about to start. Follow a similar pattern from other components.

File: controller/src/main/java/io/pravega/controller/server/health/EventProcessorHealthContributor.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+import io.pravega.controller.server.eventProcessor.ControllerEventProcessors;
+import io.pravega.shared.health.Health;
+import io.pravega.shared.health.Status;
+import io.pravega.shared.health.impl.AbstractHealthContributor;
+
+public class EventProcessorHealthContributor extends AbstractHealthContributor {
+    private final ControllerEventProcessors controllerEventProcessors;
+    public EventProcessorHealthContributor(String name, ControllerEventProcessors  controllerEventProcessors) {
+        super(name);
+        this.controllerEventProcessors =  controllerEventProcessors;
```
Comment (shrids) [inline controller/src/main/java/io/pravega/controller/server/health/EventProcessorHealthContributor.java:27]: @NotNull check here and for the below classes as well...

File: controller/src/test/java/io/pravega/controller/server/bucket/ZkStoreBucketServiceTest.java [hunk 17]
```diff
@@ -162,6 +163,7 @@ public void testOwnershipOfExistingBucket() throws Exception {
         BucketManager service2 = bucketStoreFactory.createRetentionService(Duration.ofMillis(5000), stream -> CompletableFuture.completedFuture(null), executor2);
         service2.startAsync();
         service2.awaitRunning();
+        Assert.assertTrue(service2.isHealthy());
 
         Thread.sleep(10000);
         assertTrue(service2.getBucketServices().values().stream().allMatch(x -> x.getKnownStreams().size() == 2));
```
Comment (shrids) [inline controller/src/test/java/io/pravega/controller/server/bucket/ZkStoreBucketServiceTest.java:169]: assertEquals?

File: controller/src/main/java/io/pravega/controller/server/health/GRPCServerHealthContributor.java [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+
+    public GRPCServerHealthContributor(String name, GRPCServer grpcServer) {
+        super(name);
+        this.grpcServer = grpcServer;
+    }
+
+    @Override
+    public Status doHealthCheck(Health.HealthBuilder builder) throws Exception {
+        boolean running = grpcServer.isRunning();
+        Status status = Status.DOWN;
```
Comment (shshashwat) [inline controller/src/main/java/io/pravega/controller/server/health/GRPCServerHealthContributor.java:35]: nit: This can also be written like
`
Status status = running? Status.Up: Status.DOWN;

return status;
`

File: controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java [hunk 19]
```diff
@@ -353,6 +360,17 @@ protected void startUp() {
                 grpcServer.awaitRunning();
             }
 
+            HealthContributor rootContrib = healthServiceManager.getRoot();
+            SegmentContainerMonitorHealthContributor segmentContainerMonitorHC = new SegmentContainerMonitorHealthContributor("segmentContainerMonitor", monitor );
```
Comment (shrids) [inline controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java:364]: this needs to be conditionally created based on `serviceConfig.getHostMonitorConfig().isHostMonitorEnabled()`
Please verify other components as well on similar lines.

File: controller/src/test/java/io/pravega/controller/server/health/EventProcessorHealthContributorTest.java [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+        builder = Health.builder().name("eventprocessors");
+    }
+
+    @After
+    public void tearDown() {
+        contributor.close();
+    }
+
+    @Test
+    public void
```
Comment (shrids) [inline controller/src/test/java/io/pravega/controller/server/health/EventProcessorHealthContributorTest.java:103]: nit: formatting.

File: controller/src/test/java/io/pravega/controller/server/health/WatermarkingServiceHealthContibutorTest.java [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import java.time.Duration;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static org.mockito.Mockito.*;
```
Comment (shrids) [inline controller/src/test/java/io/pravega/controller/server/health/WatermarkingServiceHealthContibutorTest.java:37]: we should remove .* in import.

File: controller/src/test/java/io/pravega/controller/server/health/WatermarkingServiceHealthContibutorTest.java [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+
+    @After
+    public void tearDown() {
+        contributor.close();
+    }
+
+    @Test
+    public void testHealthCheck() throws Exception {
+        watermarkingService.startAsync();
+        TimeUnit.SECONDS.sleep(5);
```
Comment (shrids) [inline controller/src/test/java/io/pravega/controller/server/health/WatermarkingServiceHealthContibutorTest.java:71]: can we avoid sleep here ?

**Review by shrids — APPROVED** [review]
The changes look good apart from a couple of comments. Please do take a look at them.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java:145]: To enable a method to be used for testing @VisibleForTesting annotation should do the job. An API should be made public only if production code requires it to be public and not for tests

File: controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+public class ClusterListenerHealthContributor extends AbstractHealthContributor {
+    private final ControllerClusterListener controllerClusterListener;
+
+    public ClusterListenerHealthContributor(String name, ControllerClusterListener controllerClusterListener) {
+        super(name);
+        this.controllerClusterListener = Preconditions.checkNotNull(controllerClusterListener, "controllerClusterListener");
+    }
+
+    @Override
+    public Status doHealthCheck(Health.HealthBuilder builder) throws Exception {
```
Comment (pbelgundi) [inline controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java:35]: This code looks much better now...

File: controller/src/main/java/io/pravega/controller/server/health/GRPCServerHealthContributor.java (hunk 18 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/health/GRPCServerHealthContributor.java:35]: Please do not resolve conversations unless you have responded to them or resolved them to the reviewers satisfaction.

Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/health/GRPCServerHealthContributor.java:35]: @shshashwat have refactored the code here, Please let me know if it is fine

File: shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java (hunk 7 above)
Comment (anishakj) [reply shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java:119]: Added  tests

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java:145]: without making it public, i was not able to use the constructor in my test code.

File: controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java [hunk 24]
```diff
@@ -67,6 +68,26 @@ public ControllerClusterListener(final Host host, final Cluster cluster,
...
+            return ((ClusterZKImpl) cluster).isZKConnected();
+        }
+        return false;
+    }
+
+    public boolean isSweepersReady() {
+        for (int i = 0; i < sweepers.size(); i++) {
```
[the remark this answers is not in the document]
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java]: Done

File: controller/src/main/java/io/pravega/controller/fault/SegmentContainerMonitor.java [hunk 25]
```diff
@@ -41,6 +43,9 @@
     //The ZK path which is monitored for leader selection.
     private final String leaderZKPath;
 
+    private final AtomicBoolean isZKConnected = new AtomicBoolean(true);
```
[the remark this answers is not in the document]
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/fault/SegmentContainerMonitor.java]: Done

File: shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java (hunk 6 above)
Comment (anishakj) [reply shared/cluster/src/main/java/io/pravega/common/cluster/zkImpl/ClusterZKImpl.java]: Done

File: shared/cluster/src/main/java/io/pravega/common/cluster/Cluster.java [hunk 26]
```diff
@@ -60,4 +60,12 @@
      */
     public Set<Host> getClusterMembers();
 
+    /**
+     * Get the health status.
+     *
+     * @return true by default.
+     */
+    default public boolean isHealthy() {
```
Comment (pbelgundi) [inline shared/cluster/src/main/java/io/pravega/common/cluster/Cluster.java:68]: Again, lets have an abstract method instead of "default".

File: controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java (hunk 5 above)
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/health/ClusterListenerHealthContributor.java]: Done

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: controller/src/main/java/io/pravega/controller/server/health/WatermarkingServiceHealthContributor.java (hunk 9 above)
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/health/WatermarkingServiceHealthContributor.java]: Fixed

File: controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java (hunk 10 above)
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java]: Fixed

File: controller/src/main/java/io/pravega/controller/fault/SegmentContainerMonitor.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/fault/SegmentContainerMonitor.java:97]: Ok.

File: controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java [hunk 27]
```diff
@@ -67,6 +67,33 @@ public ControllerClusterListener(final Host host, final Cluster cluster,
...
+    public boolean isMetadataServiceConnected() {
+        return cluster.isHealthy();
+    }
+
+    /**
+     * Get the sweepers status.
+     *
+     * @return true if all sweepers are ready.
+     */
+    public boolean areAllSweepersReady() {
```
Comment (pbelgundi) [inline controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java:84]: Please add code to existing tests to cover these uncovered lines reported by `codecov patch`.

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java (hunk 8 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java]: cool. thanks!

**Review by shshashwat — APPROVED** [review]

File: controller/src/main/java/io/pravega/controller/store/checkpoint/CheckpointStore.java [hunk 28]
```diff
@@ -134,4 +134,13 @@ void removeReader(final String process, final String readerGroup, final String r
             throws CheckpointStoreException;
 
     Set<String> getProcesses() throws CheckpointStoreException;
+
+    /**
+     * Get the health status.
+     *
+     * @return true by default.
+     */
+    default public boolean isHealthy() {
```
[the remark this answers is not in the document]
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/store/checkpoint/CheckpointStore.java]: Done

File: shared/cluster/src/main/java/io/pravega/common/cluster/Cluster.java (hunk 26 above)
Comment (anishakj) [reply shared/cluster/src/main/java/io/pravega/common/cluster/Cluster.java]: Done

File: controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java (hunk 27 above)
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java:84]: Done

File: controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java [hunk 29]
```diff
@@ -83,6 +83,15 @@ protected void doStart() {
 
     protected abstract int getBucketCount();
 
+    /**
+     * Get the health status.
+     *
+     * @return true by default.
+     */
+    public boolean isHealthy() {
```
Comment (pbelgundi) [inline controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java:91]: This implementation is not needed an essentially is doing the same thing as default method in the interface. Please remove it.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java:145]: Okay

**Review by pbelgundi — CHANGES_REQUESTED** [review]
Mostly LGTM, except the `isHealthy()` method implementation from bucket manager abstract class needs to be removed.

File: controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java (hunk 29 above)
Comment (anishakj) [reply controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java]: Done

**Review by pbelgundi — APPROVED** [review]
LGTM

**Review by eolivelli — APPROVED** [review]
LGTM
